From a0943bd277ed9b5ecaf1c7c1ace7590b1b6da37d Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH] RuntimeException when reading a serializable object

when ResovlerActivity is started and the 3rd party app
put an extra with a bad serial object. It will throw a
RuntimeException. Actually we don't want this exception
thrown because of the app's bad extra data

diff --git a/core/java/com/android/internal/app/ResolverActivity.java b/core/java/com/android/internal/app/ResolverActivity.java
index 1e26c92..6d6cec4 100644
--- a/core/java/com/android/internal/app/ResolverActivity.java
+++ b/core/java/com/android/internal/app/ResolverActivity.java
@@ -385,7 +385,17 @@
     }
 
     protected String getReferrerPackageName() {
-        final Uri referrer = getReferrer();
+        Uri referrer = null;
+        try {
+            referrer = getReferrer();
+        } catch(RuntimeException e) {
+            if (e.getCause() instanceof ClassNotFoundException) {
+                Log.d(TAG, "Error getReferrer\n" + e);
+                return null;
+            } else {
+                throw new RuntimeException(e);
+            }
+        }
         if (referrer != null && "android-app".equals(referrer.getScheme())) {
             return referrer.getHost();
         }
